File: android-native/app/src/main/java/com/example/kpkn/domain/nutrition/DatasetKnowledgeContext.kt
package com.example.kpkn.domain.nutrition

object DatasetKnowledgeContext {
    data class ContextProfile(val count: Int, val typicalGrams: List<Double>, val typicalKcal: List<Double>, val typicalProtein: List<Double>, val typicalFats: List<Double>, val typicalCarbs: List<Double>)
    data class MacroRange(val count: Int, val kcalMin: Double, val kcalMax: Double, val kcalMedian: Double, val proteinMin: Double, val proteinMax: Double, val proteinMedian: Double, val fatsMin: Double, val fatsMax: Double, val fatsMedian: Double, val carbsMin: Double, val carbsMax: Double, val carbsMedian: Double)

    val CONTEXT_KEYWORDS: Map<String, List<String>> = mapOf(
        "CASINO" to listOf("casino", "cafeteria", "comedor", "buffet del", "menu del dia"),
        "POST_ENTRENO" to listOf("post-entreno", "post entreno", "post-entrenamiento", "recuperacion", "post-sentadillas", "post-pecho", "post-espalda", "post-pierna"),
        "POWERBUILDER" to listOf("powerbuilder", "power builder", "volumen extremo", "masa extrema"),
        "ABUELA_CHILENA" to listOf("abuela chilena", "abuela", "contundente", "plato hondo", "tazon grande", "plato rebosante"),
        "OFICINA" to listOf("oficina", "escritorio", "trabajo", "reunion", "break de oficina"),
        "ESTUDIANTE" to listOf("estudiante", "universidad", "facultad", "campus", "barato", "corto de lucas", "sobrevivencia"),
        "CONDOMINIO" to listOf("condominio", "edificio", "departamento", "casa"),
        "SNACK" to listOf("snack", "colacion", "merendola", "merienda", "tentempie", "piscolabis", "refrigerio"),
        "DESAYUNO" to listOf("desayuno", "desayunar", "manana"),
        "ALMUERZO" to listOf("almuerzo", "almorzar", "mediodia", "tarde"),
        "CENA" to listOf("cena", "cenar", "noche", "nocturno", "antes de dormir"),
    )

    val INTENSIFIER_KEYWORDS: Map<String, List<String>> = mapOf(
        "GIGANTE" to listOf("gigante", "gigantesca", "gigantesco", "enorme", "descomunal", "bestial"),
        "GENEROSO" to listOf("generoso", "generosa", "generosos", "generosas"),
        "COLMADO" to listOf("colmado", "colmada", "colmados", "colmadas", "hasta el borde", "rebosante", "rebosantes", "lleno", "llena", "repleto"),
        "GRANDE" to listOf("grande", "grandes", "gran"),
        "PEQUENO" to listOf("pequeno", "pequena", "pequenas", "pequenos", "chico", "chica", "chicos", "chicas"),
        "FINO" to listOf("fino", "fina", "finos", "finas", "delgado", "delgada", "delgados", "delgadas"),
        "GRUESO" to listOf("grueso", "gruesa", "gruesos", "gruesas", "gordo", "gorda"),
    )

    val CONTEXT_PROFILES: Map<String, ContextProfile> = mapOf(
        "CASINO" to ContextProfile(510, listOf(180.0, 200.0, 250.0, 300.0, 150.0, 100.0, 80.0, 120.0, 60.0, 90.0), listOf(350.0, 420.0, 280.0, 310.0, 380.0, 450.0, 320.0, 290.0, 360.0, 400.0), listOf(15.0, 18.0, 12.0, 20.0, 22.0, 16.0, 14.0, 19.0, 17.0, 21.0), listOf(12.0, 15.0, 10.0, 18.0, 20.0, 14.0, 11.0, 16.0, 13.0, 17.0), listOf(45.0, 55.0, 35.0, 40.0, 50.0, 60.0, 42.0, 38.0, 48.0, 52.0)),
        "POST_ENTRENO" to ContextProfile(116, listOf(200.0, 250.0, 300.0, 180.0, 220.0, 280.0, 150.0, 190.0, 240.0, 260.0), listOf(500.0, 600.0, 700.0, 450.0, 550.0, 650.0, 480.0, 520.0, 580.0, 620.0), listOf(40.0, 50.0, 60.0, 35.0, 45.0, 55.0, 38.0, 42.0, 48.0, 52.0), listOf(10.0, 15.0, 20.0, 8.0, 12.0, 18.0, 9.0, 11.0, 14.0, 16.0), listOf(50.0, 60.0, 70.0, 45.0, 55.0, 65.0, 48.0, 52.0, 58.0, 62.0)),
        "POWERBUILDER" to ContextProfile(46, listOf(300.0, 350.0, 400.0, 280.0, 320.0, 380.0, 250.0, 290.0, 340.0, 360.0), listOf(800.0, 900.0, 1000.0, 750.0, 850.0, 950.0, 780.0, 820.0, 880.0, 920.0), listOf(60.0, 70.0, 80.0, 55.0, 65.0, 75.0, 58.0, 62.0, 68.0, 72.0), listOf(25.0, 30.0, 35.0, 22.0, 28.0, 33.0, 23.0, 26.0, 29.0, 31.0), listOf(80.0, 90.0, 100.0, 75.0, 85.0, 95.0, 78.0, 82.0, 88.0, 92.0)),
        "SNACK" to ContextProfile(200, listOf(30.0, 40.0, 50.0, 25.0, 35.0, 45.0, 20.0, 28.0, 38.0, 42.0), listOf(150.0, 200.0, 250.0, 120.0, 180.0, 220.0, 130.0, 160.0, 190.0, 210.0), listOf(8.0, 10.0, 12.0, 6.0, 9.0, 11.0, 7.0, 8.0, 10.0, 11.0), listOf(5.0, 8.0, 10.0, 4.0, 7.0, 9.0, 5.0, 6.0, 8.0, 9.0), listOf(15.0, 20.0, 25.0, 12.0, 18.0, 22.0, 14.0, 16.0, 19.0, 21.0)),
        "DESAYUNO" to ContextProfile(150, listOf(200.0, 250.0, 300.0, 180.0, 220.0, 280.0, 150.0, 190.0, 240.0, 260.0), listOf(350.0, 400.0, 450.0, 300.0, 380.0, 420.0, 320.0, 360.0, 390.0, 410.0), listOf(15.0, 18.0, 22.0, 12.0, 16.0, 20.0, 13.0, 15.0, 17.0, 19.0), listOf(10.0, 12.0, 15.0, 8.0, 11.0, 14.0, 9.0, 10.0, 12.0, 13.0), listOf(40.0, 50.0, 60.0, 35.0, 45.0, 55.0, 38.0, 42.0, 48.0, 52.0)),
        "ALMUERZO" to ContextProfile(200, listOf(250.0, 300.0, 350.0, 220.0, 280.0, 320.0, 200.0, 240.0, 290.0, 310.0), listOf(500.0, 600.0, 700.0, 450.0, 550.0, 650.0, 480.0, 520.0, 580.0, 620.0), listOf(30.0, 35.0, 40.0, 25.0, 32.0, 38.0, 28.0, 30.0, 34.0, 36.0), listOf(15.0, 20.0, 25.0, 12.0, 18.0, 22.0, 14.0, 16.0, 19.0, 21.0), listOf(50.0, 60.0, 70.0, 45.0, 55.0, 65.0, 48.0, 52.0, 58.0, 62.0)),
        "CENA" to ContextProfile(120, listOf(200.0, 250.0, 300.0, 180.0, 220.0, 280.0, 150.0, 190.0, 240.0, 260.0), listOf(400.0, 450.0, 500.0, 350.0, 420.0, 480.0, 370.0, 400.0, 440.0, 460.0), listOf(25.0, 30.0, 35.0, 20.0, 28.0, 32.0, 22.0, 25.0, 29.0, 31.0), listOf(12.0, 15.0, 18.0, 10.0, 14.0, 16.0, 11.0, 12.0, 15.0, 16.0), listOf(40.0, 50.0, 60.0, 35.0, 45.0, 55.0, 38.0, 42.0, 48.0, 52.0)),
        "OFICINA" to ContextProfile(9, listOf(30.0, 40.0, 50.0, 25.0, 35.0, 45.0, 20.0, 28.0, 38.0), listOf(150.0, 200.0, 250.0, 120.0, 180.0, 220.0, 130.0, 160.0, 190.0), listOf(5.0, 8.0, 10.0, 4.0, 7.0, 9.0, 5.0, 6.0, 8.0), listOf(5.0, 8.0, 10.0, 4.0, 7.0, 9.0, 5.0, 6.0, 8.0), listOf(15.0, 20.0, 25.0, 12.0, 18.0, 22.0, 14.0, 16.0, 19.0)),
        "ESTUDIANTE" to ContextProfile(4, listOf(200.0, 250.0, 300.0, 180.0), listOf(300.0, 400.0, 500.0, 280.0), listOf(12.0, 15.0, 20.0, 10.0), listOf(8.0, 12.0, 15.0, 7.0), listOf(30.0, 40.0, 50.0, 28.0)),
        "ABUELA_CHILENA" to ContextProfile(11, listOf(300.0, 350.0, 400.0, 280.0, 320.0, 380.0, 250.0, 290.0, 340.0, 360.0), listOf(600.0, 700.0, 800.0, 550.0, 650.0, 750.0, 580.0, 620.0, 680.0, 720.0), listOf(30.0, 35.0, 40.0, 25.0, 32.0, 38.0, 28.0, 30.0, 34.0, 36.0), listOf(20.0, 25.0, 30.0, 18.0, 22.0, 28.0, 19.0, 21.0, 24.0, 26.0), listOf(60.0, 70.0, 80.0, 55.0, 65.0, 75.0, 58.0, 62.0, 68.0, 72.0)),
        "CONDOMINIO" to ContextProfile(1, listOf(250.0), listOf(400.0), listOf(15.0), listOf(12.0), listOf(40.0)),
    )

    val MACRO_RANGES: Map<String, MacroRange> = mapOf(
        "MACRO_CALC" to MacroRange(3737, 50.0, 1600.0, 465.0, 3.0, 92.0, 46.0, 0.0, 110.0, 12.0, 1.0, 190.0, 55.0),
        "DATABASE_LOOKUP" to MacroRange(3747, 10.0, 500.0, 130.0, 0.5, 35.0, 13.0, 0.0, 45.0, 2.0, 0.0, 80.0, 15.0),
        "GENERAL" to MacroRange(8722, 1.0, 1600.0, 350.0, 0.0, 92.0, 30.0, 0.0, 110.0, 10.0, 0.0, 190.0, 45.0),
        "DESCRIBE" to MacroRange(1791, 100.0, 1600.0, 550.0, 5.0, 80.0, 40.0, 5.0, 80.0, 15.0, 10.0, 150.0, 50.0),
        "QUESTION" to MacroRange(1408, 10.0, 500.0, 150.0, 0.5, 35.0, 15.0, 0.0, 45.0, 3.0, 0.0, 80.0, 18.0),
    )
}

File: ios-native/KPKNFit/KPKNFit/Domain/Nutrition/DatasetKnowledgeContext.swift
import Foundation

enum DatasetKnowledgeContext {
    struct ContextProfile {
        let count: Int
        let typicalGrams: [Double]
        let typicalKcal: [Double]
        let typicalProtein: [Double]
        let typicalFats: [Double]
        let typicalCarbs: [Double]
    }

    struct MacroRange {
        let count: Int
        let kcalMin: Double
        let kcalMax: Double
        let kcalMedian: Double
        let proteinMin: Double
        let proteinMax: Double
        let proteinMedian: Double
        let fatsMin: Double
        let fatsMax: Double
        let fatsMedian: Double
        let carbsMin: Double
        let carbsMax: Double
        let carbsMedian: Double
    }

    static let contextKeywords: [String: [String]] = [
        "CASINO": ["casino", "cafeteria", "comedor", "buffet del", "menu del dia"],
        "POST_ENTRENO": ["post-entreno", "post entreno", "post-entrenamiento", "recuperacion", "post-sentadillas", "post-pecho", "post-espalda", "post-pierna"],
        "POWERBUILDER": ["powerbuilder", "power builder", "volumen extremo", "masa extrema"],
        "ABUELA_CHILENA": ["abuela chilena", "abuela", "contundente", "plato hondo", "tazon grande", "plato rebosante"],
        "OFICINA": ["oficina", "escritorio", "trabajo", "reunion", "break de oficina"],
        "ESTUDIANTE": ["estudiante", "universidad", "facultad", "campus", "barato", "corto de lucas", "sobrevivencia"],
        "CONDOMINIO": ["condominio", "edificio", "departamento", "casa"],
        "SNACK": ["snack", "colacion", "merendola", "merienda", "tentempie", "piscolabis", "refrigerio"],
        "DESAYUNO": ["desayuno", "desayunar", "manana"],
        "ALMUERZO": ["almuerzo", "almorzar", "mediodia", "tarde"],
        "CENA": ["cena", "cenar", "noche", "nocturno", "antes de dormir"],
    ]

    static let intensifierKeywords: [String: [String]] = [
        "GIGANTE": ["gigante", "gigantesca", "gigantesco", "enorme", "descomunal", "bestial"],
        "GENEROSO": ["generoso", "generosa", "generosos", "generosas"],
        "COLMADO": ["colmado", "colmada", "colmados", "colmadas", "hasta el borde", "rebosante", "rebosantes", "lleno", "llena", "repleto"],
        "GRANDE": ["grande", "grandes", "gran"],
        "PEQUENO": ["pequeno", "pequena", "pequenas", "pequenos", "chico", "chica", "chicos", "chicas"],
        "FINO": ["fino", "fina", "finos", "finas", "delgado", "delgada", "delgados", "delgadas"],
        "GRUESO": ["grueso", "gruesa", "gruesos", "gruesas", "gordo", "gorda"],
    ]

    static let contextProfiles: [String: ContextProfile] = [
        "CASINO": ContextProfile(count: 510, typicalGrams: [180, 200, 250, 300, 150, 100, 80, 120, 60, 90], typicalKcal: [350, 420, 280, 310, 380, 450, 320, 290, 360, 400], typicalProtein: [15, 18, 12, 20, 22, 16, 14, 19, 17, 21], typicalFats: [12, 15, 10, 18, 20, 14, 11, 16, 13, 17], typicalCarbs: [45, 55, 35, 40, 50, 60, 42, 38, 48, 52]),
        "POST_ENTRENO": ContextProfile(count: 116, typicalGrams: [200, 250, 300, 180, 220, 280, 150, 190, 240, 260], typicalKcal: [500, 600, 700, 450, 550, 650, 480, 520, 580, 620], typicalProtein: [40, 50, 60, 35, 45, 55, 38, 42, 48, 52], typicalFats: [10, 15, 20, 8, 12, 18, 9, 11, 14, 16], typicalCarbs: [50, 60, 70, 45, 55, 65, 48, 52, 58, 62]),
        "POWERBUILDER": ContextProfile(count: 46, typicalGrams: [300, 350, 400, 280, 320, 380, 250, 290, 340, 360], typicalKcal: [800, 900, 1000, 750, 850, 950, 780, 820, 880, 920], typicalProtein: [60, 70, 80, 55, 65, 75, 58, 62, 68, 72], typicalFats: [25, 30, 35, 22, 28, 33, 23, 26, 29, 31], typicalCarbs: [80, 90, 100, 75, 85, 95, 78, 82, 88, 92]),
        "SNACK": ContextProfile(count: 200, typicalGrams: [30, 40, 50, 25, 35, 45, 20, 28, 38, 42], typicalKcal: [150, 200, 250, 120, 180, 220, 130, 160, 190, 210], typicalProtein: [8, 10, 12, 6, 9, 11, 7, 8, 10, 11], typicalFats: [5, 8, 10, 4, 7, 9, 5, 6, 8, 9], typicalCarbs: [15, 20, 25, 12, 18, 22, 14, 16, 19, 21]),
        "DESAYUNO": ContextProfile(count: 150, typicalGrams: [200, 250, 300, 180, 220, 280, 150, 190, 240, 260], typicalKcal: [350, 400, 450, 300, 380, 420, 320, 360, 390, 410], typicalProtein: [15, 18, 22, 12, 16, 20, 13, 15, 17, 19], typicalFats: [10, 12, 15, 8, 11, 14, 9, 10, 12, 13], typicalCarbs: [40, 50, 60, 35, 45, 55, 38, 42, 48, 52]),
        "ALMUERZO": ContextProfile(count: 200, typicalGrams: [250, 300, 350, 220, 280, 320, 200, 240, 290, 310], typicalKcal: [500, 600, 700, 450, 550, 650, 480, 520, 580, 620], typicalProtein: [30, 35, 40, 25, 32, 38, 28, 30, 34, 36], typicalFats: [15, 20, 25, 12, 18, 22, 14, 16, 19, 21], typicalCarbs: [50, 60, 70, 45, 55, 65, 48, 52, 58, 62]),
        "CENA": ContextProfile(count: 120, typicalGrams: [200, 250, 300, 180, 220, 280, 150, 190, 240, 260], typicalKcal: [400, 450, 500, 350, 420, 480, 370, 400, 440, 460], typicalProtein: [25, 30, 35, 20, 28, 32, 22, 25, 29, 31], typicalFats: [12, 15, 18, 10, 14, 16, 11, 12, 15, 16], typicalCarbs: [40, 50, 60, 35, 45, 55, 38, 42, 48, 52]),
        "OFICINA": ContextProfile(count: 9, typicalGrams: [30, 40, 50, 25, 35, 45, 20, 28, 38], typicalKcal: [150, 200, 250, 120, 180, 220, 130, 160, 190], typicalProtein: [5, 8, 10, 4, 7, 9, 5, 6, 8], typicalFats: [5, 8, 10, 4, 7, 9, 5, 6, 8], typicalCarbs: [15, 20, 25, 12, 18, 22, 14, 16, 19]),
        "ESTUDIANTE": ContextProfile(count: 4, typicalGrams: [200, 250, 300, 180], typicalKcal: [300, 400, 500, 280], typicalProtein: [12, 15, 20, 10], typicalFats: [8, 12, 15, 7], typicalCarbs: [30, 40, 50, 28]),
        "ABUELA_CHILENA": ContextProfile(count: 11, typicalGrams: [300, 350, 400, 280, 320, 380, 250, 290, 340, 360], typicalKcal: [600, 700, 800, 550, 650, 750, 580, 620, 680, 720], typicalProtein: [30, 35, 40, 25, 32, 38, 28, 30, 34, 36], typicalFats: [20, 25, 30, 18, 22, 28, 19, 21, 24, 26], typicalCarbs: [60, 70, 80, 55, 65, 75, 58, 62, 68, 72]),
        "CONDOMINIO": ContextProfile(count: 1, typicalGrams: [250], typicalKcal: [400], typicalProtein: [15], typicalFats: [12], typicalCarbs: [40]),
    ]

    static let macroRanges: [String: MacroRange] = [
        "MACRO_CALC": MacroRange(count: 3737, kcalMin: 50, kcalMax: 1600, kcalMedian: 465, proteinMin: 3, proteinMax: 92, proteinMedian: 46, fatsMin: 0, fatsMax: 110, fatsMedian: 12, carbsMin: 1, carbsMax: 190, carbsMedian: 55),
        "DATABASE_LOOKUP": MacroRange(count: 3747, kcalMin: 10, kcalMax: 500, kcalMedian: 130, proteinMin: 0.5, proteinMax: 35, proteinMedian: 13, fatsMin: 0, fatsMax: 45, fatsMedian: 2, carbsMin: 0, carbsMax: 80, carbsMedian: 15),
        "GENERAL": MacroRange(count: 8722, kcalMin: 1, kcalMax: 1600, kcalMedian: 350, proteinMin: 0, proteinMax: 92, proteinMedian: 30, fatsMin: 0, fatsMax: 110, fatsMedian: 10, carbsMin: 0, carbsMax: 190, carbsMedian: 45),
        "DESCRIBE": MacroRange(count: 1791, kcalMin: 100, kcalMax: 1600, kcalMedian: 550, proteinMin: 5, proteinMax: 80, proteinMedian: 40, fatsMin: 5, fatsMax: 80, fatsMedian: 15, carbsMin: 10, carbsMax: 150, carbsMedian: 50),
        "QUESTION": MacroRange(count: 1408, kcalMin: 10, kcalMax: 500, kcalMedian: 150, proteinMin: 0.5, proteinMax: 35, proteinMedian: 15, fatsMin: 0, fatsMax: 45, fatsMedian: 3, carbsMin: 0, carbsMax: 80, carbsMedian: 18),
    ]
}
